### DISCUSSION TITLE
Make simple luminance data available by script

### DISCUSSION THREAD

**Discussion by Hannyson** [godotengine/godot-proposals#6149] [3D]
I try to get via script the previous frame rendered, or a mipmap version of it, each frame, to analyze it and use that to adjust some of my game's parameters.
In the shaders, the last frame is available with SCREEN_TEXTURE, it's easy and fast.
But to get it via get_viewport().get_texture() it's too slow.
I created a ColorRect with a material using SCREEN_TEXTURE to show it and capture it in a SubViewport, with every combination with BackBufferCopy, SubViewportContainer, it doesn't work; no 3D or 2D object is shown in the SubViewport, like it is separated from the other objects. The BackBufferCopy object copies what is in the current Viewport/SubViewport, not the root one which I want.

**Comment by Calinou** [comment]
This requires implementing [hardware readbacks](https://github.com/godotengine/godot-proposals/issues/7886) to be done in a fast way, which has a 2-3 frame delay. There is no way to have the viewport data available on the next frame on the CPU without incurring a significant slowdown due to GPU-CPU synchronization.

**Comment by Hannyson** [comment]
If it is just a few frames delay, I really don't mind.
The only way I made it work is with creating a second Camera3D in a SubViewport, which takes 10 fps out of 30 for my game.
If it can speed things up or simplify them, a mipmap version would be enough (like SCREEN_TEXTURE in the shaders). With the slow second camera I use, I created a SubViewport as small as 256 x 144 pixels, that was ok.

**Comment by Hannyson** [comment]
Up, if it is possible to only get a simple float value indicative of luminance would be alright, if it's not already available somewhere, a value something like .auto_exposure_* you can get by script (didn't find it in the docs).
What I search is a simple value, so I can adapt my camera aperture myself by script rather than changing automatically the exposure.

I gave a brief look at the servers/rendering/renderer_rd/renderer_scene_render_rd.cpp where the auto-exposure is launched from (I think), it uses luminance buffers. Is there a way to access that ?

**Reply by Calinou** [reply]
> Is there a way to access that ?

No, these buffers are not accessible from shaders. [More low-level shader access is planned in the future](https://github.com/godotengine/godot/pull/80214), but it's intended for edge cases and may not resolve your particular use case.

**Reply by Hannyson** [reply]
And on the idea of a simple float value like Camera / WorldEnvironment .auto_exposure_current_luminance ?

**Reply by Calinou** [reply]
I believe the exposure value is computed on the GPU exclusively, so the CPU is completely unaware of the current autoexposure value.
